### PR TITLE
fix(explore): deprecated x periods pattern in new time picker value

### LINKED ISSUE #12508: [Explore] New time picker breaks old charts with "x days" in time picker start date
### Expected results

In 0.38 and earlier, if you set "30 days" or "5 months" in start date (without "ago", "before", or "last"), it will be automatically converted to the equivalent of "30 days ago". The new time picker will treat it as "30 days after today", causing an `Unexpected time range: From date cannot be larger than to date` error for some existing charts. 

Here are examples of time ranges that will break in the new time picker but are valid for the old one:

1. "30 days: "  -> "30 days ago : now"
3. "7 days: 7 days" -> "7 days ago : 7 days after today"

### Actual results

Existing charts with "x periods" set in start date should behave the same.

Since the original behavior is a little bit ambiguous, we may actually want to force users to use "30 days ago" or "- 30 days" for this use case. So I think it's OK to keep current logic, but to minimize disruption, we should do a db migration to automatically fix such charts and add a breaking change warning in the `v1.0` changelog (albeit numbers users who used this syntax is low).

### Environment

Latest master

### PR DESCRIPTION
**Pull request by zhaoyongjie**
### SUMMARY
closes: #12508 

1. changed "No filter" show in the date-picker control pill.
2. deprecated x periods, if user input "x [dateunit][s]" response an error.
3. raise input error when the user input as an arbitrary string.
4. added [x periods] -> [x periods ago/later] DB migration script

#### 1. "No filter" show in the date-picker control pill
image

#### 2. deprecated x periods
image

#### 3. catch arbitrary input
image

### TEST PLAN
UT covered these case

### ADDITIONAL INFORMATION

- [x] Has associated issue: #12508 
- [ ] Changes UI
- [x] Requires DB Migration.
- [x] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Review by junlincc — COMMENTED** [review]
https://user-images.githubusercontent.com/67837651/104822308-00ab6e80-57ee-11eb-934c-bbdc95f8f0e6.mov

Thanks for the PR! the error msg is not quite human readable or actionable still... @mihir174  we need your help here.

**Comment by ktmud** [comment]
```
Start time too ambiguous. Do you mean "30 days ago" or "now + 30 days"?
```

**Comment by junlincc** [comment]
thanks for the fix! @zhaoyongjie, all changes LGTM! 

Question: when i type '3' in the box, the modal populates 2021-01-03, with 3 as date. 
vs. when i type in '40', modal takes it as year... what's the logic behind? should we consider both as unclear input and prompt users to be more specific before generating a time range as well?  

<img width="933" alt="Screen Shot 2021-01-17 at 6 14 48 PM" src="https://user-images.githubusercontent.com/67837651/104868489-f7a2c600-58ef-11eb-9673-d6c3d93c1fc1.png"

**Review by villebro — COMMENTED** [review]
I agree with the change, but this change will break existing charts relying on the new logic. This is a chart that was working before but now is failing with the new validation:
image
Since it seems `dateutil.parser.parse` interprets "7 days" as interchangeable with "7 days ago", we should perhaps do a migration that adds the "ago" suffix to old time ranges to avoid breaking them? Alternatively we should potentially only do the validation for new charts, and let old charts keep working with the deprecated ranges.

**Comment by ktmud** [comment]
I'll vote for a db migration as initially proposed in #12508

**Comment by zhaoyongjie** [comment]
> thanks for the fix! @zhaoyongjie, all changes LGTM!
> 
> Question: when i type '3' in the box, the modal populates 2021-01-03, with 3 as date.
> vs. when i type in '40', modal takes it as year... what's the logic behind? should we consider both as unclear input and prompt users to be more specific before generating a time range as well?
> 
> Screen Shot 2021-01-17 at 6 14 48 PM

Hi @junlincc 

This is a good point. This behaviour looks like this: when the input string is less than 31, it will be derived as the date, when it is greater than 31 it will be derived as the year. This is `parsedatetime.Calendar()` behavior.
https://github.com/apache/superset/blob/84266bfa9ed1acd532eea611ec322a1eaa9480fa/superset/utils/date_parser.py#L78-L79

If the datetime derived(parse) is to be improved. It is need to specify a fixed format for date parse.

**Comment by zhaoyongjie** [comment]
> I agree with the change, but this change will break existing charts relying on the new logic. This is a chart that was working before but now is failing with the new validation:
> image
> Since it seems `dateutil.parser.parse` interprets "7 days" as interchangeable with "7 days ago", we should perhaps do a migration that adds the "ago" suffix to old time ranges to avoid breaking them? Alternatively we should potentially only do the validation for new charts, and let old charts keep working with the deprecated ranges.

I'm concerned that a db metadata with a large number of charts will require some migration risk to do this migration.

Are we considering compatibility with the old syntax?  I would prefer to be compatible with the original syntax.

What do you think about this? @ktmud @villebro

**Comment by villebro** [comment]
> I'm concerned that a db metadata with a large number of charts will require some migration risk to do this migration.
> 
> Are we considering compatibility with the old syntax? I would prefer to be compatible with the original syntax.

It's true that migrations are risky, and migrations should be avoided unless required. However, similar migrations have been done before, and I'm pretty confident we can write a migration that handles this correctly, as it's not that complex. At least we should be able to fix the vast majority of cases that would otherwise break with just a simple regex. The reason I personally prefer to either support the old format or create a migration is because:
- For SIP-15 there was a deprecation warning that urged people to edit the chart and save it after checking that the new time range endpoints were ok. See e.g. https://github.com/apache/superset/pull/11417 . Point being we should try to avoid breaking existing charts unless it is unavoidable, especially post 1.0.
- People tend to learn to learn to use shorthand whenever supported, and if `7 days` has historically worked equally well as `7 days ago`, I would suspect many charts out there will be using the old notation. This is probably also highly skewed - some users might not even notice, while other's might have always use the deprecated syntax, hence breaking lots of charts for certain users.

If it's not difficult to implement, I would prefer to let old charts work, but force the user to update the time range if they edit the chart. But if that's difficult to implement or gets messy, I'd prefer to do a migration.

**Comment by ktmud** [comment]
I don't think the migration is particularly risky. You can pre-select the list of charts that use the invalid format of date string and update them one by one (for Airbnb, it's only a couple of hundreds such charts among about 16 million).

The query is as simple as a list of RegExp matches:

```sql
SELECT
  id,
  params
FROM "slices"
WHERE regexp_like(params, '"time_range": "\d+ (days|weeks|months|years|quarters)')
   AND not regexp_like(params, '"time_range": "\d+.*(ago|before).*:')
```

For MySQL and SQLite, you can use `.op('regexp')(...)`, for PostgreSQL, you can use `.op('similar to')(...)`.

**Comment by ktmud** [comment]
@zhaoyongjie can we also add debounce to the backend verification API? The API should probably also return 200 instead of 400 when verification failed (400 should be reserved for malformed/invalid payload).

**Comment by zhaoyongjie** [comment]
@ktmud , I will open a new PR to improve that.

**Review by ktmud — APPROVED** [review]

File: superset/utils/date_parser.py [hunk 1]
```diff
@@ -71,19 +71,32 @@ def parse_human_datetime(human_readable: str) -> datetime:
...
+    x_periods = r"^\s*([0-9]+)\s+(second|minute|hour|day|week|month|quarter|year)s?\s*$"
+    if re.search(x_periods, human_readable, re.IGNORECASE):
+        raise ValueError(
+            _(
+                "date string is unclear."
+                " Please specify [{0} ago] or [{0} later]".format(human_readable)
+            )
+        )
+
+    error_msg = ValueError(_("Couldn't parse date string [{}]".format(human_readable)))
```
Comment (ktmud) [inline superset/utils/date_parser.py:83]: Proper i18n should use the built-in interpolation from `_`:

```python
_("Couldn't parse date string [%(input)s]", input="human_readable")
```

Comment (zhaoyongjie) [reply superset/utils/date_parser.py:83]: Thanks for pointing this out. I remember that `lazy_gettext()` function can support this syntax, but I've tried this and now it doesn't work.  I will fix this.

File: superset/migrations/versions/260bf0649a77_migrate_x_dateunit_in_time_range.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+def upgrade():
+    bind = op.get_bind()
+    session = db.Session(bind=bind)
+    x_dateunit_in_since = DateRangeMigration.x_dateunit_in_since
+    x_dateunit_in_until = DateRangeMigration.x_dateunit_in_until
+
+    if isinstance(bind.dialect, SQLiteDialect):
+        try:
+            # The REGEXP operator is a special syntax for the regexp() user function.
```
Comment (zhaoyongjie) [inline superset/migrations/versions/260bf0649a77_migrate_x_dateunit_in_time_range.py:63]: SQLite does not support REGEXP syntax by default.
https://www.sqlite.org/lang_expr.html#regexp

File: superset/migrations/versions/260bf0649a77_migrate_x_dateunit_in_time_range.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+
+def upgrade():
+    bind = op.get_bind()
+    session = db.Session(bind=bind)
+    x_dateunit_in_since = DateRangeMigration.x_dateunit_in_since
+    x_dateunit_in_until = DateRangeMigration.x_dateunit_in_until
+
+    if isinstance(bind.dialect, SQLiteDialect):
+        try:
```
Comment (zhaoyongjie) [inline superset/migrations/versions/260bf0649a77_migrate_x_dateunit_in_time_range.py:62]: SQLite does not support REGEXP syntax by default.
https://www.sqlite.org/lang_expr.html#regexp

File: superset/migrations/versions/260bf0649a77_migrate_x_dateunit_in_time_range.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+            # https://www.sqlite.org/lang_expr.html#regexp
+            to_lower = sa.func.LOWER
+            where_clause = or_(
+                sa.func.REGEXP(to_lower(Slice.params), x_dateunit_in_since),
+                sa.func.REGEXP(to_lower(Slice.params), x_dateunit_in_until),
+            )
+        except OperationalError:
+            return
+
+    if isinstance(bind.dialect, MySQLDialect):
```
Comment (zhaoyongjie) [inline superset/migrations/versions/260bf0649a77_migrate_x_dateunit_in_time_range.py:73]: MySQL 5.7 support REGEXP operator, the REGEXP_LIKE function is not supported until MySQL 8.x
https://dev.mysql.com/doc/refman/5.7/en/regexp.html#operator_regexp

File: superset/utils/date_parser.py [hunk 5]
```diff
@@ -71,19 +71,35 @@ def parse_human_datetime(human_readable: str) -> datetime:
     >>> year_after_1 == year_after_2
     True
     """
+    x_periods = r"^\s*([0-9]+)\s+(second|minute|hour|day|week|month|quarter|year)s?\s*$"
+    if re.search(x_periods, human_readable, re.IGNORECASE):
+        raise ValueError(
+            _(
+                "date string is unclear."
```
Comment (ktmud) [inline superset/utils/date_parser.py:78]:
```suggestion
                "Date string is unclear."
```

**Review by ktmud — APPROVED** [review]
LGTM with one more nit

**Comment by betodealmeida** [comment]
@zhaoyongjie I'm writing the notes for 1.1 and saw this PR. I just wanted to clarify something for the notes: this PR doesn't **deprecate** the usage of `x periods`, it **removes** it, right?

Looking at the screenshots it seems that the user cannot input "30 days", since "apply" is disabled. If this were a deprecation the user would see the warning, but would still be able to use the deprecated format, until we eventually removed it in the future.

**Comment by zhaoyongjie** [comment]
@betodealmeida , The user is no longer allowed to enter `x periods`

This issue migrates `x [periods]` to `x [periods] last/ago` in database

**Pull request merged**